### ISSUE TITLE
when work-0 is in 6 epoch，work-1 have be in 7 . how to shuffled data before per epoch ？

### ISSUE THREAD

**Issue by zira-wang** [horovod/horovod#1456]
**Environment:**
1. Framework: TensorFlow
2. Framework version: 1.5.0
3. Horovod version: 0.16.0
4. MPI version: openmpi/4.0.0
5. CUDA version: cuda/9.0 
6. NCCL version:  nccl/2.1.15
7. Python version: 3.6.0
8. OS and version: linux
9. GCC version: 

**Checklist:**
1. Did you search issues to find if somebody asked this question before?
2. If your question is about hang, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/running.rst)?
3. If your question is about docker, did you read [this doc](https://github.com/horovod/horovod/blob/master/docs/docker.rst)?
4. Did you check if you question is answered in the [troubleshooting guide](https://github.com/horovod/horovod/blob/master/docs/troubleshooting.rst)?

**Your question:**
Please ask your question here.

I want to shuf the data before per epoch。 My code：
==============
        for step in range(FLAGS['epoch']):
            if hvd.local_rank() == 0:
                os.system('bash shuf.sh ' + os.path.join(FLAGS['train_data'],"seq.in") + " " + os.path.join(FLAGS['train_data'],"seq.out"))
==============
The data have be shuffled， in this way。But，when work-0 is in 6 epoch，work-1 have be in 7 echo. Can't horovod Synchronous epoch data ？

**Comment by tgaddair** [comment]
Hey @zira-wang, let me make sure I understand your use case.  It looks like you're performing a shuffle on disk on worker 0 between each epoch, but while worker 0 is doing the shuffle, the other workers are proceeding ahead to the next epoch, is that right?

One thing I would suggest you do to workaround this is to make use of a barrier operation following the shuffle so that the other workers do not try and get ahead of worker 0 while it's shuffling:

```
hvd.allreduce([], name="Barrier")
```

To be clear, it should not happen that different workers have successfully been able to process a different number of epochs.  Because Horovod is a synchronous distributed training framework, workers will stall and wait for one another to catch up before proceeding to the next batch.  If you're seeing otherwise, it means that Horovod is not correctly setup in your training script to aggregate gradients across workers.

**Comment by zira-wang** [comment]
> Hey @zira-wang, let me make sure I understand your use case. It looks like you're performing a shuffle on disk on worker 0 between each epoch, but while worker 0 is doing the shuffle, the other workers are proceeding ahead to the next epoch, is that right?
> 
> One thing I would suggest you do to workaround this is to make use of a barrier operation following the shuffle so that the other workers do not try and get ahead of worker 0 while it's shuffling:
> 
> ```
> hvd.allreduce([], name="Barrier")
> ```
> 
> To be clear, it should not happen that different workers have successfully been able to process a different number of epochs. Because Horovod is a synchronous distributed training framework, workers will stall and wait for one another to catch up before proceeding to the next batch. If you're seeing otherwise, it means that Horovod is not correctly setup in your training script to aggregate gradients across workers.

Thank u. My bug has been solved。My bug has been solved. The reason for this error is that I did not replace "tf.gradients" with "optimizer.compute_gradients" .

**Comment by zira-wang** [comment]
> Hey @zira-wang, let me make sure I understand your use case. It looks like you're performing a shuffle on disk on worker 0 between each epoch, but while worker 0 is doing the shuffle, the other workers are proceeding ahead to the next epoch, is that right?
> 
> One thing I would suggest you do to workaround this is to make use of a barrier operation following the shuffle so that the other workers do not try and get ahead of worker 0 while it's shuffling:
> 
> ```
> hvd.allreduce([], name="Barrier")
> ```
> 
> To be clear, it should not happen that different workers have successfully been able to process a different number of epochs. Because Horovod is a synchronous distributed training framework, workers will stall and wait for one another to catch up before proceeding to the next batch. If you're seeing otherwise, it means that Horovod is not correctly setup in your training script to aggregate gradients across workers.
